src/desktop/widgets/rulerwidget.cpp

```diff
@@ -0,0 +1,162 @@
+
+#include "desktop/widgets/rulerwidget.h"
+
+#include <QPainter>
+
+#include <array>
+
+namespace widgets {
+static constexpr std::array niceDivisions{1,   2,	5,	 10,   20,	 50,
+										  100, 200, 500, 1000, 2000, 5000};
+
+
+RulerWidget::RulerWidget(QWidget *parent)
+	: QWidget(parent)
+{
+	// This widget does not yet support creating guides, so for now don't
+	// intercept mouse events to avoid confusion.
+	setAttribute(Qt::WA_TransparentForMouseEvents);
+}
+
+void RulerWidget::paintEvent(QPaintEvent *)
+{
+	QPainter painter(this);
+	painter.fillRect(0, 0, width(), height(), palette().color(QPalette::Light));
+
+	// The ruler doesn't make any sense right now for a rotated canvas, so
+	// disable it if the canvas is rotated.
+	// TODO: We could project the pixel grid to the ruler and draw ticks at the
+	//       intersection.
+	if(m_isRotated)
+		return;
+
+	// How many canvas pixels are in view.
+	qreal distance = m_lMax - m_lMin;
+	if(distance < 1)
+		return;
+
+	// We want about 60 display independent pixels per division.
+	int divisions = longSize() / 60;
+
+	qreal initialPPD = distance / divisions;
+	// Having weird numbers for divisions isn't great, so try to fit this to a
+	// nearby nice number.
+	int adjustedPPD = 1;
+	qreal bestDist = niceDivisions.back();
+	for(int ppd : niceDivisions) {
+		qreal dist = std::abs(initialPPD - ppd);
+		if(dist < bestDist) {
+			adjustedPPD = ppd;
+			bestDist = dist;
+		}
+	}
+
+	int pixelsPerRulerDivision =
+		std::lround((adjustedPPD * m_canvasToRulerScale));
+	divisions = longSize() / pixelsPerRulerDivision;
+
+	QColor dark = palette().color(QPalette::Dark);
+	QColor text = palette().color(QPalette::Text);
+
+	painter.setPen(dark);
+	painter.setRenderHint(QPainter::Antialiasing, true);
+	// Set pen for consistent line width on high DPI displays
+	QPen pen = painter.pen();
+	pen.setWidthF(1.0 / devicePixelRatioF());
+	painter.setPen(pen);
+
+	qreal startOffset = std::fmod(std::abs(m_lMin), adjustedPPD);
+	int canvasL = std::lround(
+		m_lMin < 0 ? (m_lMin + startOffset) - adjustedPPD
+				   : m_lMin - startOffset);
+
+	int lineHeight = 0;
+	if(!isHorizontal()) {
+		// Figure out the largest number we're going to draw so that we can
+		// choose a smaller font size if it's too big.
+		auto digitsNeeded = [&](int num) {
+			int negative = 0;
+			if(num < 0) {
+				num = -num;
+				negative = 1;
+			}
+			for(int i = 1; i <= std::numeric_limits<int>::digits10; ++i) {
+				if(num < std::pow(10, i))
+					return i + negative;
+			}
+			return std::numeric_limits<int>::digits10 + 1 + negative;
+		};
+
+		int digits = std::max(
+			digitsNeeded(canvasL),
+			digitsNeeded(canvasL + adjustedPPD * divisions));
+
+		while(true) {
+			lineHeight =
+				painter.fontMetrics().tightBoundingRect("0123456789").height();
+			lineHeight += 2;
+
+			if(lineHeight * digits >= pixelsPerRulerDivision - 2) {
+				QFont smallerFont = painter.font();
+				int newPointSize = smallerFont.pointSize() - 1;
```

Comment:
QFont inexplicably has two incompatible ways of specifying its size, either as points or as pixels. Asking it for the "wrong" kind of size will give you -1, see <https://doc.qt.io/qt-5/qfont.html#pixelSize>.

You could handle that if you want, although other UI elements of Drawpile don't do it either. The only system where this seems to happen in practice is Android, where the font size is ridiculous to begin with and we override it with a point size by default, so the cases where you get a pixel font are probably pathological.

Reply:
I'll adjust the check so it bails out in that case too. Probably better to use the default size than to try and call setPointSize with a negative number.